### PR TITLE
bump msal

### PR DESCRIPTION
**Pull request by derisen**
## Purpose

* Bumped msal-browser to 2.13.1
* Added new combined SUSI user flow with password reset (B2C_1_susi_reset_v2)
* Added new msal.js logout APIs (logoutPopup, logoutRedirect)
* Added a simple filter for choosing the correct account when there are multiple accounts in cache due to initiating a user-flow like edit profile (ps. this can be improved)

## Does this introduce a breaking change?

```
[ ] Yes
[ x ] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[ x ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ x ] Other... Please describe: package update
```

## How to Test
*  Get the code

```
git clone https://github.com/Azure-Samples/ms-identity-b2c-javascript-spa.git
cd ms-identity-b2c-javascript-spa
git checkout derisen
npm install
```

## What to Check
Verify that the following are valid
1) sign-in
2) call API
3) edit profile
4) reset password and repeat 2, 3

## Other Information

Strangely, when returning from password reset, initiating edit profile user-flow prompts a sign-in screen. This does not happen if you initiate edit-profile after a normal sign-in. I'm suspecting that there is a difference between the session cookie issued after normal sign-in and after returning from password reset. @nickgmicrosoft @tnorling any insights on this?

### CODE REVIEW

File: App/authPopup.js [hunk 1]
```diff
@@ -14,11 +14,21 @@ function selectAccount () {
 
     const currentAccounts = myMSALObj.getAllAccounts();
 
-    if (currentAccounts.length === 0) {
+    if (!currentAccounts || currentAccounts.length < 1) {
```
Comment (tnorling) [inline App/authPopup.js]: `getAllAccounts` should return an empty array if no accounts are found. Was there a reason you had to add a check for truthiness here?

File: App/authPopup.js [hunk 2]
```diff
@@ -14,11 +14,21 @@ function selectAccount () {
...
-        console.log("Multiple accounts detected.");
+    } else if (currentAccounts.length === Object.entries(b2cPolicies.names).length) {
+        
+        /**
+         * Due to the way MSAL caches account objects, the auth response from initiating a user-flow
+         * is cached as a new account. Here we make sure we are selecting the account with homeAccountId
+         * that contains the sign-up/sign-in user-flow.
+         */
+        const account = currentAccounts.find(account => 
+            account.homeAccountId.toUpperCase().includes(b2cPolicies.names.signUpSignIn.toUpperCase()));
```
Comment (tnorling) [inline App/authPopup.js]: Ah, nice idea! If you want to go this route I would also recommend checking `clientId` and authority match and then also that there aren't more than 1 user signed into this authority/policy.

Another way to do this would be to track the `homeAccountId` after retrieving tokens and use the `getAccountByHomeId` API or use the `setActiveAccount`/`getActiveAccount` APIs.

File: App/authPopup.js [hunk 3]
```diff
@@ -55,16 +65,6 @@ function signIn() {
         .then(handleResponse)
         .catch(error => {
             console.log(error);
-                
-            // Error handling
-            if (error.errorMessage) {
-                // Check for forgot password error
-                // Learn more about AAD error codes at https://docs.microsoft.com/en-us/azure/active-directory/develop/reference-aadsts-error-codes
-                if (error.errorMessage.indexOf("AADB2C90118") > -1) {
```
Comment (tnorling) [inline App/authPopup.js:63]: Curious if all new sign-in/sign-up policies will have the built-in password reset by default? @nickgmicrosoft 

Would it make sense to continue demonstrating how to catch the error and call a dedicated password reset policy just for customers that are still using separate policies?

File: App/authPopup.js (hunk 2 above)
Comment (tnorling) [reply App/authPopup.js]: Either way, this is a great improvement over logging out and logging back in again!

File: App/authRedirect.js [hunk 4]
```diff
@@ -7,19 +7,20 @@ let username = "";
...
-        // Learn more about AAD error codes at https://docs.microsoft.com/en-us/azure/active-directory/develop/reference-aadsts-error-codes
-        if (error.errorMessage.indexOf("AADB2C90118") > -1) {
-            try {
-                myMSALObj.loginRedirect(b2cPolicies.authorities.forgotPassword);
-            } catch(err) {
-                console.log(err);
+    .then(response => {
+        if (response) {
+            /**
+             * We need to reject id tokens that were not issued with the default sign-in policy.
```
Comment (tnorling) [inline App/authRedirect.js]: Maybe reword this? You're not exactly rejecting the idToken, just only updating the UI if a user signed-in, correct?

File: App/authPopup.js [hunk 5]
```diff
@@ -75,13 +75,14 @@ function signOut() {
      * https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-browser/docs/request-response-object.md#request
      */
 
-    // Choose which account to logout from.
-    
     const logoutRequest = {
-        account: myMSALObj.getAccountByHomeId(accountId)
+        postLogoutRedirectUri: msalConfig.auth.redirectUri,
+        redirectMainWindowTo: msalConfig.auth.redirectUri
```
Comment (tnorling) [inline App/authPopup.js]:
```suggestion
        mainWindowRedirectUri: msalConfig.auth.redirectUri
```

File: App/authPopup.js [hunk 6]
```diff
@@ -75,13 +75,14 @@ function signOut() {
...
     };
     
-    myMSALObj.logout(logoutRequest);
+    myMSALObj.logoutPopup(logoutRequest).then(() => {
+        window.location.reload();
```
Comment (tnorling) [inline App/authPopup.js]: This may not be needed if you're redirecting the main window with the request

File: App/authPopup.js (hunk 1 above)
Comment (derisen) [reply App/authPopup.js]: Oops, I was trying to align in it with our docs. Will revert this and update the [doc](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-browser/docs/accounts.md)

File: App/authPopup.js (hunk 3 above)
Comment (derisen) [reply App/authPopup.js:63]: I'll defer to you and @nickgmicrosoft here. Worst case I can keep one of the b2c samples the old way.

File: App/authPopup.js (hunk 6 above)
Comment (derisen) [reply App/authPopup.js]: reloading the page? that's what I expected but the main window stayed the same after logout with popup. Will double check.

Comment (derisen) [reply App/authPopup.js]: my bad, it wasn't working 'cause I had the property name wrong in `logoutRequest`. removing..

File: App/authPopup.js (hunk 2 above)
Comment (sameerag) [reply App/authPopup.js]: @derisen Is this a part of the improved B2C policy experience or an enhancement we added?

Comment (derisen) [reply App/authPopup.js]: @tnorling thanks a lot for the help! I was under the impression that `set/getActiveAccount` is Angular specific. To keep track of the `homeAccountId`, would it be safe to use the browser storage? I wasn't sure to go that route. Failing that, I've come up with something like this:

```javascript
function selectAccount() {
    /**
     * See here for more info on account retrieval: 
     * https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-common/docs/Accounts.md
     */

    const currentAccounts = myMSALObj.getAllAccounts();

    if (currentAccounts.length < 1) {
        return;
    } else if (currentAccounts.length > 1) {

        /**
         * Due to the way MSAL caches account objects, the auth response from initiating a user-flow
         * is cached as a new account. Here we make sure we are selecting the account with homeAccountId
         * that contains the sign-up/sign-in user-flow, as this is the default flow we initially signed-in with.
         */
        const accounts = currentAccounts.filter(account =>
            account.homeAccountId.toUpperCase().includes(b2cPolicies.names.signUpSignIn.toUpperCase())
            &&
            account.idTokenClaims.iss.toUpperCase().includes(b2cPolicies.authorityDomain.toUpperCase())
            &&
            account.idTokenClaims.aud === msalConfig.auth.clientId 
            );

        if (accounts.length > 1) {
            if (accounts.every(account => account.localAccountId === accounts[0].localAccountId)) {
                // All accounts belong to the same user
                setAccount(accounts[0]);
            } else {
                // Multiple users detected
                signOut();
            };
        } else if (accounts.length === 1) {
            setAccount(accounts[0]);
        }

    } else if (currentAccounts.length === 1) {
        setAccount(currentAccounts[0]);
    }
}
```

@sameerag this is just an enhancement. Forcing to sign out the user wasn't the best experience.

**Review by sameerag — APPROVED** [review]
Approving. One comment for future: Can we have a sample that includes a new policy merge b2c did for SISU and password edit? Nick shared the details.
